lib/data/hash.ml

```diff
@@ -0,0 +1,25 @@
+open import "../prelude.ml"
+
+class eq 'a => hashable 'a
+  val hash : 'a -> int
+
+instance hashable int
+  let hash x = x
+
+instance hashable ()
+  let hash () = 0
+
+instance hashable bool
+  let hash x = if x then 1 else 0
+
+instance hashable 'a * hashable 'b => hashable ('a * 'b)
+  let hash (a, b) = 31 * hash a + hash b
+
+instance hashable string
+  let hash xs =
+    (* Hash using the djb2 hash function *)
+    let open import "../lua/string.ml"
+    let rec go x = function
+    | 0 -> x
```

Comment:
Shouldn't we take the remainder of this by a suitable prime?

Reply:
Moreover, this algorithm works to hash any Foldable, does it not? We should at least add instances for lists.